util/src/main/java/io/camunda/zeebe/util/startup/StartupProcess.java

```diff
@@ -0,0 +1,190 @@
+      logger.info("Finished startup process");
+    } else {
+      final var stepToStart = stepsToStart.remove(0);
+      startedSteps.push(stepToStart);
+
+      logCurrentStep("Startup", stepToStart);
+      stepToStart
+          .startup(context)
+          .whenComplete(
+              (contextReturnedByStep, error) -> {
+                if (error != null) {
+                  logger.warn("Aborting startup process due to exception.", error);
+                  startupFuture.completeExceptionally(error);
+                } else {
+                  proceedWithStartup(stepsToStart, contextReturnedByStep, startupFuture);
+                }
+              });
+    }
+  }
+
+  /**
+   * Executes the shutdown logic
+   *
+   * @param context the shutdown context at the start of this step
+   * @return future with the shutdown context at the end of this step.
+   */
+  public synchronized CompletableFuture<CONTEXT> shutdown(final CONTEXT context) {
+    logger.debug("Shutdown was called with context: " + context);
+    if (!startupCalled) {
+      throw new IllegalStateException("shutdown(...) can only be called after startup(...)");
+    }
+    if (shutdownFuture == null) {
+      logger.info("Starting shutdown process");
+      shutdownFuture = new CompletableFuture<>();
+
+      proceedWithShutdown(context, shutdownFuture, new ArrayList<>());
+    } else {
+      logger.info("Shutdown already in progress");
+    }
+    return shutdownFuture;
+  }
+
+  private void proceedWithShutdown(
+      final CONTEXT context,
+      final CompletableFuture<CONTEXT> shutdownFuture,
+      final List<Throwable> collectedExceptions) {
+    if (startedSteps.isEmpty()) {
+      completeShutdownFuture(context, shutdownFuture, collectedExceptions);
+    } else {
+      final var stepToShutdown = startedSteps.pop();
+
+      logCurrentStep("Shutdown", stepToShutdown);
+      stepToShutdown
+          .shutdown(context)
+          .whenComplete(
+              (contextReturnedByShutdown, error) -> {
+                final CONTEXT contextToUse;
+                if (error != null) {
+                  collectedExceptions.add(error);
+                  contextToUse = context;
+                } else {
+                  contextToUse = contextReturnedByShutdown;
+                }
+
+                proceedWithShutdown(contextToUse, shutdownFuture, collectedExceptions);
+              });
+    }
+  }
+
+  private void completeShutdownFuture(
+      final CONTEXT context,
+      final CompletableFuture<CONTEXT> shutdownFuture,
+      final List<Throwable> collectedExceptions) {
+    if (collectedExceptions.isEmpty()) {
+      shutdownFuture.complete(context);
+      logger.info("Finished shutdown process");
+    } else {
+      final Throwable exception;
+      if (collectedExceptions.size() == 1) {
+        exception = collectedExceptions.get(0);
+      } else {
+        exception = new Exception("Exceptions occurred during shutdown");
```

Comment:
:x: One thing I'm missing here is knowing which stage failed - we collect exceptions, and I get the exceptions, but I don't know which stage failed. I suppose I can infer it from the stack trace, but it would be great to know quickly which are the stages that produced exceptions. So we could here write something like (just a suggestion, feel free to improve):

> Exceptions occurred during shutdown: [ stage X: ConnectException, stage Y: TimeoutException ]. See the suppressed exceptions for more.